walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,94 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    Set<String> attendees =  new HashSet<String>();
+    int meetingDur = (int) request.getDuration();
+    attendees.addAll(request.getAttendees());
+    if (meetingDur > TimeRange.WHOLE_DAY.duration()){
+      return possibleTimes;
+    }
+    if (events.isEmpty() || attendees.isEmpty()){
+      possibleTimes.add(TimeRange.WHOLE_DAY);
+      return possibleTimes;
+    }
+    ArrayList<TimeRange> busyWithOverlap = getBusyTimes(events, attendees);
+    possibleTimes = findFreeTimes(busyWithOverlap, request.getDuration());
+    return possibleTimes;
   }
-}
+
+/**
+  * Returns an ArrayList of meeting times that are unavailable due to schedule
+  * conflicts and overlapping meeting times within the given group of attendees. 
+  *
+  * @param  events the Collection of Event objects passed into the query function
+  * @param  requestAttendees the Set of attendees from the MeetingRequest parameter in query
+  * @return    the ArrayList of all meeting times that do not work with the attendees schedules
+  */
+  private ArrayList<TimeRange> getBusyTimes(Collection<Event> events, Set<String> requestAttendees) {
+    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    for(Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      Set<String> overlap = new HashSet<>(eventAttendees);
+      overlap.retainAll(requestAttendees);
+      if (!overlap.isEmpty()){
+        busyTimes.add(event.getWhen());
+      }
+    }
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    ArrayList<TimeRange> busyWithOverlap = new ArrayList<TimeRange>();
+    if(busyTimes.size() > 0) {
+      busyWithOverlap.add(busyTimes.get(0));
+    }
+    int index = 0;
```

Comment:
use something like busyWithOverlayIndex and busyTimesIndex, instead of index and i to make it clear which belongs to which

Reply:
Good idea, changed in last PR